### PR TITLE
feat(TreeView): allow selection without expansion

### LINKED ISSUE #7378: Tree view - ability to select parent nodes without changing expand state
Placeholder until the core work is complete.

Core issue - https://github.com/patternfly/patternfly/issues/4724

### PR DESCRIPTION
**Pull request by kmcfaul**
**What**: Closes #7378

Adds `hasSelectableNodes` prop that separates expansion & selection behaviors - clicking the row will select it while clicking the toggle button will trigger the expansion. This flag will also allow parent nodes to be selected (in addition to leaf nodes) and apply the current style modifier to them.

### CODE REVIEW

**Comment by kmcfaul** [comment]
@mcoker One thing I noticed on the core PR was that the `pf-c-tree-view__node` was still a `button` (and the toggle was in a `span`) in the `With selectable, expandable nodes` example, which I think was the example showcasing this behavioral change, but in your PR description you mentioned that the node should now be a `div` with the toggle changing to a `button`. I went with the `div > button` structure right now but let me know if I misunderstood and should change something.

**Comment by mcoker** [comment]
@kmcfaul sorry for the confusion, we made another update after I opened the PR and I didn't update the description. What's in the PR URL, or maybe more accurately the core workspace. Can you go from that? Or would you prefer I go through the changes and update the PR description?

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx [hunk 1]
```diff
@@ -94,14 +97,23 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
   }, [isExpanded, defaultExpanded]);
 
   const Component = hasCheck ? 'div' : 'button';
```
Comment (thatblindgeye) [inline packages/react-core/src/components/TreeView/TreeViewListItem.tsx:99]: I was able to replicate the behavior from the "with checkboxes" example by updating this to `hasCheck || hasSelectableNodes ? 'div' : 'button'` (more info in another comment below)

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx [hunk 2]
```diff
@@ -94,14 +97,23 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
...
-  const ToggleComponent = hasCheck ? 'button' : 'div';
+  let ToggleComponent: 'div' | 'span' | 'button' = 'div';
+  if (hasSelectableNodes) {
+    ToggleComponent = 'span';
+  }
+  if (hasCheck) {
+    ToggleComponent = 'button';
+  }
```
Comment (thatblindgeye) [inline packages/react-core/src/components/TreeView/TreeViewListItem.tsx:106]: ...and having the toggle set to a button when either of these props are passed in.

Plain keyboard functionality would still require some tweaking to get working properly, though.

**Review by thatblindgeye — COMMENTED** [review]
I'm noticing issues when trying to navigate/interact with the example via VO or keyboard. Essentially I can select a tree item fine, but I can't find a way to toggle the expand state in anyway. See the following video (using VO):

https://user-images.githubusercontent.com/70952936/180041380-8848c3e1-ae18-4eaf-80ce-89a3cd52c30a.mov

The "with checkboxes" example works a bit better in terms of VO/keyboard being able to expand/collapse a tree item or check/uncheck the checkbox without one affecting the other. I made some comments below about what I tried doing locally to try and resolve this.

**Review by mcoker — CHANGES_REQUESTED** [review]
tl;dr - Looking at this PR, looks like we need to:
* Make `.pf-c-tree-v-ew__node-container` a `span`
* Make `.pf-c-tree-view__node-toggle` a `span` in expandable, non-checkbox nodes
* Looks like all nodes of a selectable tree view are getting `.pf-m-selectable` - only the expandable node needs it.
* Make checkbox nodes a `label`, and change the node-text element inside to a `span`
* Add `.pf-m-selectable` to the expandable checkbox nodes

I also found a few bugs and put up a core PR - https://github.com/patternfly/patternfly/issues/4990

----

Reviewing the core PR, the changes are:

* A tree view node has `.pf-m-selectable` when it's expandabale/selectable (including checkbox nodes)
* These elements changed to a `<span>` for HTML validity reasons, not a big deal:
  * `.pf-c-tree-view__node-container`
  * `.pf-c-tree-view__node-content`
  * `.pf-c-tree-view__node-toggle` when it isn't a button
* Checkbox `.pf-c-tree-view__node` is now a `<label>`
  * `.pf-c-tree-view__node-text` is a `<span>` since it was a `<label>`
  * HTML label attrs moved from node-text to node
* Updated hover states on the toggle icon so that it changes color when you're hovering the thing that will toggle expansion. ie, the node if non-selectable, the toggle if selectable.

With the intention that basically
 * A default tree view node is a `button`, the toggle is a `span`. The whole node toggles expansion, the toggle icon is just for show and changes colors when you hover/focus on the node.
* A checkbox node is a `label`, and the toggle inside is a `button`. Clicking on the toggle will toggle expansion, clicking anywhere else will trigger the checkbox to be checked/selected.
* A non-checkbox selectable node is a `button` and the toggle is a `span` (since you can't nest a button in a button). Clicking the toggle toggles the node expansion, clicking anywhere else selects the node.

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx (hunk 2 above)
Comment (kmcfaul) [reply packages/react-core/src/components/TreeView/TreeViewListItem.tsx:106]: I had both of these component types set up the way you are describing, but the structure didn't match core so I backed them out. It looks like there's a few more updates that this PR needs to fully match core (below) so hopefully that will iron out the behavior a bit.

**Comment by mmenestr** [comment]
This looks good overall, my only question is whether we should allow someone to select a full expandable node in a tree view - or whether they should only be allowed to select a child node with no expansion?

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx (hunk 2 above)
Comment (mcoker) [reply packages/react-core/src/components/TreeView/TreeViewListItem.tsx]: @kmcfaul updated the core structure so that selectable/expandable nodes are now a `div` with a `tabindex` so you can focus them, and the toggle inside is now a `button`. https://github.com/patternfly/patternfly/pull/4991

You should be able to make that change in your PR without pulling in core, there is no CSS or anything needed to support that change.

**Comment by kmcfaul** [comment]
@mmenestr That is the default behavior if you are referring to the blue bold styling that is applied to the selected node. I adjusted it to apply to a parent node because the new flag allows a parent node to be selected directly now, but I can change it back if we don't want that visually. LMK

**Comment by thatblindgeye** [comment]
I opened #7759 as a followup for the keyboard nav issues

**Comment by mmenestr** [comment]
Ok, no no if we're allowing a parent node to be selected then it should have that styling! Looks good!

**Review by mmenestr — APPROVED** [review]

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx [hunk 3]
```diff
@@ -186,20 +198,21 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
...
                   onSelect && onSelect(evt, itemData, parentItem);
-                  if (children && evt.isDefaultPrevented() !== true) {
+                  if (!hasSelectableNodes && children && evt.isDefaultPrevented() !== true) {
                     setIsExpanded(!internalIsExpanded);
                   }
                 }
               }}
-              tabIndex={-1}
+              tabIndex={hasSelectableNodes ? 1 : -1}
+              {...((hasCheck || hasSelectableNodes) && { htmlFor: randomId, id: `label-${randomId}` })}
```
Comment (mcoker) [inline packages/react-core/src/components/TreeView/TreeViewListItem.tsx:207]: `for` should only apply to `hasCheck` - ie, when the node is a `<label>`.

The `id` is good here, but technically doesn't need to apply to non-expandable selectable nodes. I think the case would be `hasCheck || (hasSelectableNodes && children)` to apply an `id`

**Review by mcoker — CHANGES_REQUESTED** [review]
LGTM other than the keyboard issue @thatblindgeye noted and one small nit on the `for` and `id` attrs on nodes.

I did rework the aria-labels on the nodes and toggles in this PR if we want to consider that - https://github.com/patternfly/patternfly/pull/4991

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx (hunk 3 above)
Comment (kmcfaul) [reply packages/react-core/src/components/TreeView/TreeViewListItem.tsx]: Updated. There will be a follow up to iron out the remaining keyboard issues.

**Review by mcoker — APPROVED** [review]
LGTM! I'd like to get https://github.com/patternfly/patternfly/pull/4991 merged and pulled in so we can make sure it fixes the toggle button's icon color when that node is selected.

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx [hunk 4]
```diff
@@ -186,20 +198,22 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
...
                 if (!hasCheck) {
                   onSelect && onSelect(evt, itemData, parentItem);
-                  if (children && evt.isDefaultPrevented() !== true) {
+                  if (!hasSelectableNodes && children && evt.isDefaultPrevented() !== true) {
                     setIsExpanded(!internalIsExpanded);
                   }
                 }
               }}
-              tabIndex={-1}
+              tabIndex={hasSelectableNodes ? 1 : -1}
```
Comment (thatblindgeye) [inline packages/react-core/src/components/TreeView/TreeViewListItem.tsx:206]:
```suggestion
              tabIndex={hasSelectableNodes ? 0 : -1}
```

Though it looks like this tab index is being overridden anyway by TreeViewRoot (line 38).

File: packages/react-core/src/components/TreeView/TreeViewListItem.tsx [hunk 5]
```diff
@@ -186,20 +198,22 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
...
+              {...(hasCheck && { htmlFor: randomId })}
+              {...((hasCheck || (hasSelectableNodes && children)) && { id: `label-${randomId}` })}
```
Comment (thatblindgeye) [inline packages/react-core/src/components/TreeView/TreeViewListItem.tsx:208]: The ID that gets set here includes "checkbox" in the value regardless of whether there's a checkbox or not. May be worth updating the randomId prefix to render "checkbox" or "selectable" or something depending on whether `hasSelectableNodes` is passed in.

File: packages/react-core/src/components/TreeView/TreeView.tsx [hunk 6]
```diff
@@ -45,6 +45,8 @@ export interface TreeViewProps {
   hasBadges?: boolean;
   /** Flag indicating if tree view has guide lines. */
   hasGuides?: boolean;
+  /** Flag indicating that tree nodes should be independently selectable, even when having children */
+  hasSelectableNodes?: boolean;
```
Comment (thatblindgeye) [inline packages/react-core/src/components/TreeView/TreeView.tsx:49]: Nitpick that I wouldn't block over: `isSelectable` could be an alternative prop name, as it would also align with our Card and Table selectable variants that have a similar prop.

**Review by thatblindgeye — CHANGES_REQUESTED** [review]
Just a few comments below, let me know what you think, but otherwise this looks good!

File: packages/react-core/src/components/TreeView/TreeView.tsx (hunk 6 above)
Comment (kmcfaul) [reply packages/react-core/src/components/TreeView/TreeView.tsx:49]: Yeah the prop name is something I wasn't too sure on. I thought about `isSelectable` but wasn't sure if it would be clear that it would make the nodes selectable opposed to the whole tree. I'm open to either prop name if anyone feels strongly one way or the other.

Comment (thatblindgeye) [reply packages/react-core/src/components/TreeView/TreeView.tsx:49]: Ah that's a fair point. I might think that having selectable items would make the tree view selectable in nature, but I do see how it isn't exactly clear. How do you feel about the idea of `TreeView` with prop `hasSelectableNodes`, and the `TreeViewListItem` with prop `isSelectable`?

Comment (kmcfaul) [reply packages/react-core/src/components/TreeView/TreeView.tsx:49]: I like that, will update

**Review by thatblindgeye — APPROVED** [review]

**Comment by mcoker** [comment]
@kmcfaul https://github.com/patternfly/patternfly/pull/4991 has been merged and is in core v4.206.1

**Review by tlabaj — APPROVED** [review]
lgtm
